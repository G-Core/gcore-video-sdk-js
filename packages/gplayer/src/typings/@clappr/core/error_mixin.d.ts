import '@clappr/core';
// import ErrorMixin from "@clappr/core/src/base/error_mixin/error_mixin";

// export as namespace "@clappr/core";
declare module "@clappr/core" {
    // export ErrorMixin;

    type ErrorOptions = {
        useCodePrefix: boolean;
    }

    declare interface ErrorMixin {
        createError(error: Object, options?: ErrorOptions): Error;
    }
}
