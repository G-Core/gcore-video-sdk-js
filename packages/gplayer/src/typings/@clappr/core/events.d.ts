// // Events.Custom
// import { Events } from "@clappr/core";
// declare module "@clappr/core" {
//     declare class Events {
//         static Custom: Record<string, string>;
//     }
// }