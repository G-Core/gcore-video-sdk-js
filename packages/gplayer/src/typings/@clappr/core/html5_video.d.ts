import '@clappr/core';

declare module "@clappr/core" {
    declare type HTML5VideoSettingsItem =
        | "seekbar"
        | "playpause"
        | "position"
        | "duration"
        | "fullscreen"
        | "volume"
        | "hd-indicator"
        | "playstop";

    declare type HTML5VideoSettings = {
        default: HTML5VideoSettingsItem[];
        left?: HTML5VideoSettingsItem[];
        right?: HTML5VideoSettingsItem[];
        seekEnabled?: boolean;
    }

    declare class HTML5Video extends Playback {
        settings: HTML5VideoSettings;

        readonly _minDvrSize: number;

        _onDurationChange(): void;
    }
}
