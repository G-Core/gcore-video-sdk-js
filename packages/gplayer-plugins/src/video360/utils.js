// check if the browser supports cors
export const corsSupport = (function() {
  const video = document.createElement('video');

  video.crossOrigin = 'anonymous';

  return video.hasAttribute('crossorigin');
})();

export const validProjections = [
  '360',
  '360_LR',
  '360_TB',
  '360_CUBE',
  'EAC',
  'EAC_LR',
  'NONE',
  'AUTO',
  'Sphere',
  'Cube',
  'equirectangular',
  '180'
];

export const getInternalProjectionName = function(projection) {
  if (!projection) {
    return;
  }

  projection = projection.toString().trim();

  if ((/sphere/i).test(projection)) {
    return '360';
  }

  if ((/cube/i).test(projection)) {
    return '360_CUBE';
  }

  if ((/equirectangular/i).test(projection)) {
    return '360';
  }

  for (let i = 0; i < validProjections.length; i++) {
    if (new RegExp('^' + validProjections[i] + '$', 'i').test(projection)) {
      return validProjections[i];
    }
  }
};
