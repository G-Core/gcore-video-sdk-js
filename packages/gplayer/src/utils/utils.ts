// import LogManager from './LogManager';
import { Browser } from '@clappr/core';
import assert from 'assert';

export function getLocation(href: string) {
  const l = document.createElement('a');

  l.href = href;

  return l;
}

export function strtimeToMiliseconds(str: string): number {
  if (!str) {
    return 0;
  }
  const arr = str.split(/:/);
  let h = 0, m = 0, s = 0;

  if (arr.length >= 3) {
    h = parseInt(arr[arr.length - 3]) * 60 * 60;
  } else {
    h = 0;
  }
  if (arr.length >= 2) {
    m = parseInt(arr[arr.length - 2]) * 60;
  } else {
    m = 0;
  }

  if (arr.length >= 1) {
    s = parseInt(arr[arr.length - 1]);
  } else {
    s = 0;
  }

  return (h + m + s);
}

// TODO refactor
export function isFullscreen(el: HTMLElement): boolean {
  const video = el.nodeName === "video" ? el as HTMLVideoElement : el.querySelector('video');
  assert.ok(video, 'element must be a video or contain a video element');

  if (Browser.isiOS) {
    return FullscreenIOS.isFullscreen(video);
  }
  return !!(document.fullscreenElement);
}

export const FullscreenIOS = {
  isFullscreen: function (el: HTMLVideoElement): boolean {
    try {
      if (el.webkitDisplayingFullscreen !== undefined) {
        return !!(el.webkitDisplayingFullscreen);
      }
    } catch (e) {
      // LogManager.exception(error);
      reportError(e);
    }

    return false;
  }
};
