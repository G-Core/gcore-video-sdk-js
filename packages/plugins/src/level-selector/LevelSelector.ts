import { Events, template, UICorePlugin, $ } from '@clappr/core';
import { trace } from '@gcorevideo/player';

import { CLAPPR_VERSION } from '../build.js';
import { QualityLevelInfo, ZeptoResult } from '../types.js';

import buttonHtml from '../../assets/level-selector/button.ejs';
import listHtml from '../../assets/level-selector/list.ejs';
import hdIcon from '../../assets/icons/new/hd.svg';
import arrowRightIcon from '../../assets/icons/new/arrow-right.svg';
import arrowLeftIcon from '../../assets/icons/new/arrow-left.svg';
import checkIcon from '../../assets/icons/new/check.svg';
import '../../assets/level-selector/style.scss';


const VERSION = '0.0.1';

const T = 'plugins.level_selector';

type PlaybackLevelInfo = {
  level: {
    bitrate: number;
    width: number;
    height: number;
  };
  id: number;
}

const AUTO = -1;

const NO_LEVEL: PlaybackLevelInfo = {
  level: {
    bitrate: 0,
    width: 0,
    height: 0,
  },
  id: -1,
};

export class LevelSelector extends UICorePlugin {
  private currentLevel: PlaybackLevelInfo | null = null;

  private levels: PlaybackLevelInfo[] = [];

  private levelLabels: string[] = [];

  private removeAuto = false;

  get name() {
    return 'level_selector';
  }

  get supportedVersion() {
    return { min: CLAPPR_VERSION };
  }

  static get version() {
    return VERSION;
  }

  get attributes() {
    return {
      'class': this.name,
      'data-level-selector': ''
    };
  }

  private currentText = 'Auto';

  private selectedLevelId = -1;

  //   constructor(core) {
  //     super(core);
  //   }

  get events() {
    return {
      'click .gear-sub-menu_btn': 'onLevelSelect',
      'click .gear-option': 'onShowLevelSelectMenu',
      'click .go-back': 'goBack',
    };
  }

  bindEvents() {
    this.listenTo(this.core, Events.CORE_READY, this.bindPlaybackEvents);
    this.listenTo(this.core.mediaControl, Events.MEDIACONTROL_CONTAINERCHANGED, this.reload);
    // this.listenTo(this.core.mediaControl, Events.MEDIACONTROL_RENDERED, this.render);
    this.listenTo(this.core, 'gear:rendered', this.render);
    // this.listenTo(this.core.mediaControl, Events.MEDIACONTROL_HIDE, this.hideSelectLevelMenu);
  }

  unBindEvents() {
    this.stopListening(this.core, Events.CORE_READY, this.bindPlaybackEvents);
    this.stopListening(this.core.mediaControl, Events.MEDIACONTROL_CONTAINERCHANGED, this.reload);
    // this.stopListening(this.core.mediaControl, Events.MEDIACONTROL_RENDERED);
    this.stopListening(this.core, 'gear:rendered', this.render);
    // this.stopListening(this.core.mediaControl, Events.MEDIACONTROL_HIDE);
  }

  private bindPlaybackEvents() {
    // this.currentLevel = {};
    this.currentLevel = NO_LEVEL;
    this.removeAuto = false;
    const currentPlayback = this.core.activePlayback;

    this.listenTo(currentPlayback, Events.PLAYBACK_LEVELS_AVAILABLE, this.fillLevels);
    this.listenTo(currentPlayback, Events.PLAYBACK_LEVEL_SWITCH_START, this.startLevelSwitch);
    this.listenTo(currentPlayback, Events.PLAYBACK_LEVEL_SWITCH_END, this.stopLevelSwitch);
    this.listenTo(currentPlayback, Events.PLAYBACK_BITRATE, this.updateCurrentLevel);

    this.listenTo(currentPlayback, Events.PLAYBACK_STOP, this.onStop);

    const playbackLevelsAvailableWasTriggered = currentPlayback.levels && currentPlayback.levels.length > 0;

    playbackLevelsAvailableWasTriggered && this.fillLevels(currentPlayback.levels);
  }

  onStop() {
    const currentPlayback = this.core.activePlayback;

    this.listenToOnce(currentPlayback, Events.PLAYBACK_PLAY, () => {
      trace(`${T} PLAYBACK_PLAY`, {
        playbackType: currentPlayback.getPlaybackType(),
      });
      if (currentPlayback.getPlaybackType() === 'live') {
        if (this.selectedLevelId !== -1) {
          currentPlayback.currentLevel = this.findLevelBy(this.selectedLevelId) || NO_LEVEL;
        }
      }
    });
  }

  reload() {
    this.unBindEvents();
    this.bindEvents();
    this.bindPlaybackEvents();
  }

  shouldRender() {
    if (!this.core.activeContainer) {
      return false;
    }

    const currentPlayback = this.core.activePlayback;

    if (!currentPlayback) {
      return false;
    }

    // TODO typeof currentPlayback.currentLevel === 'number' should it be
    const respondsToCurrentLevel = currentPlayback.currentLevel !== undefined;
    // Only care if we have at least 2 to choose from
    const hasLevels = !!(this.levels && this.levels.length > 1);

    return respondsToCurrentLevel && hasLevels;
  }

  render() {
    if (this.shouldRender()) {
      // this.$el.html(this.template({ 'levels': this.levels, 'title': this.getTitle() }));
      const t = template(buttonHtml);

      this.$el.html(t({
        currentText: this.currentText,
        hdIcon,
        arrowRightIcon,
      }));

      this.core.mediaControl.$el?.find('.gear-options-list [data-quality]').html(this.el);

      // this.highlightCurrentLevel();
    }

    if (this.removeAuto) {
      this.core.mediaControl
        .$el
        .find('.gear-options-list [data-quality]')
        .find('[data-level-selector-select="-1"]')
        .parent()
        .remove();
    }

    return this;
  }

  private fillLevels(levels: PlaybackLevelInfo[], initialLevel = AUTO) {
    trace('${T} fillLevels', { levels, initialLevel });
    // Player.player.trigger('levels', levels);
    // this.core.trigger('levels', levels);
    // TODO fire directly on the plugin object
    // Remove quality selector if it's not HLS
    if (initialLevel !== -1) {
      this.removeAuto = true;
    }

    // if (this.selectedLevelId === undefined) { // TODO compare with AUTO?
    //   this.selectedLevelId = initialLevel;
    // }
    this.levels = levels;
    this.configureLevelsLabels();
    this.render();
  }

  configureLevelsLabels() {
    trace('${T} configureLevelsLabels', { options: this.core.options, levels: this.levels });

    const labels = this.core.options.levelSelector?.labels;
    this.levelLabels = [];

    if (labels) {
      for (let i = 0; i < this.levels.length; i++) {
        const level = this.levels[i];
        const ll = level.level.width > level.level.height ? level.level.height : level.level.width;
        const label = labels[ll] || `${ll}p`;
        this.levelLabels.push(label);
      }
    }

    trace('${T} configureLevelsLabels leave', { levelLabels: this.levelLabels });

  }

  private findLevelBy(id: number): PlaybackLevelInfo | undefined {
    return this.levels.find((level) => level.id === id);
  }

  private onLevelSelect(event: MouseEvent) {
    const selectedLevel = (event.currentTarget as HTMLElement)?.dataset?.id;
    trace('${T} onLevelSelect', { selectedLevel });
    this.setIndexLevel(parseInt((event.currentTarget as HTMLElement)?.dataset?.id ?? "-1", 10));
    // this.toggleContextMenu();
    // event.stopPropagation();

    // return false;
  }

  goBack() {
    this.core.trigger('gear:refresh');
  }

  setIndexLevel(index: number) {
    this.selectedLevelId = index;
    if (this.core.activePlayback) {
      this.core.activePlayback.trigger('playback:level:select:start');
    }
    if (this.core.activePlayback.currentLevel.id === this.selectedLevelId) {
      return false;
    }
    this.core.activePlayback.currentLevel = this.selectedLevelId;

    try {
      this.updateText(this.selectedLevelId);
      this.highlightCurrentLevel();
    } catch (error) {
      // LogManager.exception(error);
      reportError(error);
    }
  }

  onShowLevelSelectMenu() {
    const t = template(listHtml);

    this.$el.html(t({
      levels: this.levels,
      labels: this.levelLabels,
      arrowLeftIcon,
      checkIcon,
    }));

    this.core.mediaControl.$el?.find('.gear-wrapper').html(this.el);
    this.highlightCurrentLevel();
  }

  allLevelElements() {
    return this.$('ul.gear-sub-menu li') as ZeptoResult;
  }

  levelElement(id = -1) {
    return (this.$(`ul.gear-sub-menu a[data-id="${id}"]`) as ZeptoResult).parent();
  }

  getTitle() {
    return (this.core.options.levelSelector || {}).title;
  }

  startLevelSwitch() {
    // Player.player.trigger('startLevelSwitch');
    this.levelElement(this.selectedLevelId).addClass('changing');
  }

  stopLevelSwitch() {
    if (this.core.activePlayback) {
      this.core.activePlayback.trigger('playback:level:select:end',
        {
          label: this.getLevelLabel(this.selectedLevelId)
        });
    }
    this.levelElement(this.selectedLevelId).removeClass('changing');
    // Player.player.trigger('stopLevelSwitch');
  }

  private updateText(level: number) {
    if (level === undefined || isNaN(level)) {
      return;
    }
    this.currentText = this.getLevelLabel(level);
  }

  private getLevelLabel(id: number): string {
    if (id === -1) {
      return 'Auto';
    }
    const index = this.levels.findIndex((l) => l.id === id);
    if (index < 0) {
      return 'Auto';
    }
    return this.levelLabels[index] || formatLevelLabel(this.levels[index]);
  }

  private updateCurrentLevel(info: QualityLevelInfo) {
    trace('${T} updateCurrentLevel', { info, levels: this.levels });
    const level = this.findLevelBy(info.level);

    this.currentLevel = level ? level : null;
    this.highlightCurrentLevel();
    // Player.player.trigger('updateCurrentLevel', info);
  }

  private highlightCurrentLevel() {
    trace('${T} highlightCurrentLevel', { currentLevel: this.currentLevel, selectedLevelId: this.selectedLevelId });
    this.allLevelElements().removeClass('current');
    this.allLevelElements().find('a').removeClass('gcore-skin-active');

    if (this.currentLevel) {
      const currentLevelElement = this.levelElement(this.selectedLevelId);

      currentLevelElement.addClass('current');
      currentLevelElement.find('a').addClass('gcore-skin-active');
    }

    this.updateText(this.selectedLevelId);
  }
}

function formatLevelLabel(level: PlaybackLevelInfo): string {
  const h = level.level.width > level.level.height ? level.level.height : level.level.width;
  return `${h}p`;
}