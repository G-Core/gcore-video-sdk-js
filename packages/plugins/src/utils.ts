import { Browser } from '@clappr/core';
import { reportError } from '@gcorevideo/player';

export function getLocation(href: string) {
  const l = document.createElement('a');

  l.href = href;

  return l;
}

export function strtimeToMiliseconds(str: string): number {
  if (!str) {
    return 0;
  }
  const arr = str.split(/:/);
  let h = 0, m = 0, s = 0;

  if (arr.length >= 3) {
    h = parseInt(arr[arr.length - 3]) * 60 * 60;
  } else {
    h = 0;
  }
  if (arr.length >= 2) {
    m = parseInt(arr[arr.length - 2]) * 60;
  } else {
    m = 0;
  }

  if (arr.length >= 1) {
    s = parseInt(arr[arr.length - 1]);
  } else {
    s = 0;
  }

  return (h + m + s);
}

// TODO refactor
export function isFullscreen(el: HTMLElement): boolean {
  const video = el.nodeName === "video" ? el as HTMLVideoElement : el.querySelector('video');
  if (!video) {
    return false;
  }
  if (Browser.isiOS) {
    return FullscreenIOS.isFullscreen(video);
  }
  return !!(document.fullscreenElement);
}

const FullscreenIOS = {
  isFullscreen: function (el: HTMLVideoElement): boolean {
    try {
      // @ts-ignore
      if (el.webkitDisplayingFullscreen !== undefined) {
        // @ts-ignore
        return !!(el.webkitDisplayingFullscreen);
      }
    } catch (e) {
      reportError(e);
    }

    return false;
  }
};

export function getPageX(event: MouseEvent | TouchEvent): number {
  if ((event as MouseEvent).pageX) {
    return (event as MouseEvent).pageX;
  }

  if ((event as TouchEvent).changedTouches) {
    return (event as TouchEvent).changedTouches[(event as TouchEvent).changedTouches.length - 1].pageX;
  }

  return 0;
}
