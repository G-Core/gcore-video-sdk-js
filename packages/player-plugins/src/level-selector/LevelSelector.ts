import { Events, template, UICorePlugin } from '@clappr/core'
import { type QualityLevel } from '@gcorevideo/player'
import { reportError, trace } from '@gcorevideo/utils'
import { CLAPPR_VERSION } from '../build.js'
import { ZeptoResult } from '../types.js'

import buttonHtml from '../../assets/level-selector/button.ejs'
import listHtml from '../../assets/level-selector/list.ejs'
import hdIcon from '../../assets/icons/new/hd.svg'
import arrowRightIcon from '../../assets/icons/new/arrow-right.svg'
import arrowLeftIcon from '../../assets/icons/new/arrow-left.svg'
import checkIcon from '../../assets/icons/new/check.svg'
import '../../assets/level-selector/style.scss'

import pkg from '../../package.json' with { type: 'json' }

const T = 'plugins.level_selector'
const VERSION = pkg.versions.level_selector

type TemplateFunction = (data: Record<string, unknown>) => string

/**
 * Allows to control the quality level of the playback.
 * @beta
 *
 * @remarks
 * The plugin is rendered as a button in the gear menu.
 * When clicked, it shows a list of quality levels to choose from.
 *
 * Configuration options:
 *
 * - `labels`: The labels to show in the level selector. [vertical resolution]: string
 * - `restrictResolution`: The maximum resolution to allow in the level selector.
 * - `title`: The title to show in the level selector.
 *
 * @example
 * ```ts
 * {
 *   levelSelector: {
 *     restrictResolution: 360,
 *     labels: { 360: '360p', 720: '720p' },
 *   },
 * }
 * ```
 */
export class LevelSelector extends UICorePlugin {
  private levels: QualityLevel[] = []

  private levelLabels: string[] = []

  private removeAuto = false

  private isHd = false

  private isOpen = false

  private buttonTemplate: TemplateFunction | null = null

  private listTemplate: TemplateFunction | null = null

  get name() {
    return 'level_selector'
  }

  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  static get version() {
    return VERSION
  }

  override get attributes() {
    return {
      class: this.name,
      'data-level-selector': '',
    }
  }

  private currentText = 'Auto'

  private selectedLevelId = -1

  override get events() {
    return {
      'click .gear-sub-menu_btn': 'onLevelSelect',
      'click .gear-option': 'onShowLevelSelectMenu',
      'click .go-back': 'goBack',
    }
  }

  override bindEvents() {
    this.listenTo(this.core, Events.CORE_READY, () => this.bindPlaybackEvents())
    this.listenTo(this.core, Events.CORE_ACTIVE_CONTAINER_CHANGED, () => this.bindPlaybackEvents())
    this.listenTo(this.core, 'gear:rendered', this.render)
  }

  private unBindEvents() {
    this.stopListening(this.core, Events.CORE_READY, () =>
      this.bindPlaybackEvents(),
    )
    this.stopListening(
      this.core.mediaControl,
      Events.MEDIACONTROL_CONTAINERCHANGED,
      this.reload,
    )
    this.stopListening(this.core, 'gear:rendered', this.render)
  }

  private bindPlaybackEvents() {
    this.removeAuto = false
    this.isHd = false

    const activePlayback = this.core.activePlayback

    this.listenTo(activePlayback, Events.PLAYBACK_LEVELS_AVAILABLE, (levels: QualityLevel[]) =>
      this.fillLevels(levels),
    )
    this.listenTo(
      activePlayback,
      Events.PLAYBACK_LEVEL_SWITCH_START,
      this.onLevelSwitchStart,
    )
    this.listenTo(
      activePlayback,
      Events.PLAYBACK_LEVEL_SWITCH_END,
      this.onLevelSwitchEnd,
    )
    this.listenTo(
      activePlayback,
      Events.PLAYBACK_BITRATE,
      this.updateCurrentLevel,
    )
    this.listenTo(activePlayback, Events.PLAYBACK_STOP, this.onStop)
    this.listenTo(
      activePlayback,
      Events.PLAYBACK_HIGHDEFINITIONUPDATE,
      (isHd: boolean) => {
        this.isHd = isHd
        this.deferRender()
      },
    )
    if (activePlayback?.levels?.length > 0) {
      this.fillLevels(activePlayback.levels)
    }
  }

  private onStop() {
    trace(`${T} onStop`)
    const currentPlayback = this.core.activePlayback

    this.listenToOnce(currentPlayback, Events.PLAYBACK_PLAY, () => {
      trace(`${T} on PLAYBACK_PLAY after stop`, { selectedLevelId: this.selectedLevelId })
      if (currentPlayback.getPlaybackType() === 'live') {
        if (this.selectedLevelId !== -1) {
          currentPlayback.currentLevel = this.selectedLevelId
        }
      }
    })
  }

  private reload() {
    this.unBindEvents()
    this.bindEvents()
    this.bindPlaybackEvents()
  }

  private shouldRender() {
    if (!this.core.activeContainer) {
      return false
    }

    const activePlayback = this.core.activePlayback

    if (!activePlayback) {
      return false
    }

    const supportsCurrentLevel = 'currentLevel' in activePlayback
    if (!supportsCurrentLevel) {
      return false
    }
    // Only care if we have at least 2 to choose from
    return !!(this.levels && this.levels.length > 1)
  }

  override render() {
    if (!this.shouldRender()) {
      return this
    }

    this.renderButton()

    return this
  }

  private renderButton() {
    if (!this.buttonTemplate) {
      this.buttonTemplate = template(buttonHtml)
    }

    if (!this.isOpen) {
      const html = this.buttonTemplate?.({
        arrowRightIcon,
        currentText: this.currentText,
        isHd: this.isHd,
        hdIcon,
      })
      this.$el.html(html)
      this.core.mediaControl.$el
        ?.find('.gear-options-list [data-quality]')
        .html(this.el)
    }
  }

  private renderDropdown() {
    if (!this.listTemplate) {
      this.listTemplate = template(listHtml)
    }
    const html = this.listTemplate!({
      arrowLeftIcon,
      checkIcon,
      labels: this.levelLabels,
      levels: this.levels,
      maxLevel: this.maxLevel,
      removeAuto: this.removeAuto,
    })
    this.$el.html(html)
    this.core.mediaControl.$el?.find('.gear-wrapper').html(this.el)
  }

  private get maxLevel() {
    const maxRes = this.core.options.levelSelector?.restrictResolution
    return maxRes
      ? this.levels.findIndex(
          (level) =>
            level.height === maxRes,
        )
      : -1
  }

  private fillLevels(levels: QualityLevel[]) {
    const maxResolution = this.core.options.levelSelector?.restrictResolution
    this.levels = levels
    this.makeLevelsLabels()
    if (maxResolution) {
      this.removeAuto = true
      // TODO account for vertical resolutions, i.e, normalize the resolution to the width
      const initialLevel = levels
        .filter((level) => level.height <= maxResolution)
        .pop()
      this.setLevel(initialLevel?.level ?? 0)
    }
    this.deferRender()
  }

  private makeLevelsLabels() {
    const labels = this.core.options.levelSelector?.labels ?? {}
    this.levelLabels = []

    for (let i = 0; i < this.levels.length; i++) {
      const level = this.levels[i]
      const ll = level.width > level.height ? level.height : level.width
      const label = labels[ll] || `${ll}p`
      this.levelLabels.push(label)
    }
  }

  private findLevelBy(id: number): QualityLevel | undefined {
    return this.levels.find((level) => level.level === id)
  }

  private onLevelSelect(event: MouseEvent) {
    const selectedLevel = parseInt(
      (event.currentTarget as HTMLElement)?.dataset?.id ?? '-1',
      10,
    )
    this.setLevel(selectedLevel)
    event.stopPropagation()
    return false
  }

  private goBack() {
    trace(`${T} goBack`)
    this.isOpen = false
    this.core.trigger('gear:refresh')
    this.deferRender()
  }

  private setLevel(index: number) {
    trace(`${T} setIndexLevel`, { index })
    this.selectedLevelId = index
    if (!this.core.activePlayback) {
      return
    }
    if (this.core.activePlayback.currentLevel === this.selectedLevelId) {
      return
    }
    this.core.activePlayback.currentLevel = this.selectedLevelId

    try {
      this.highlightCurrentLevel()
    } catch (error) {
      reportError(error)
    }
    this.deferRender()
  }

  private onShowLevelSelectMenu() {
    trace(`${T} onShowLevelSelectMenu`)
    this.isOpen = true
    this.renderDropdown()
    this.highlightCurrentLevel()
  }

  private allLevelElements() {
    return this.$('ul.gear-sub-menu li') as ZeptoResult
  }

  private levelElement(id = -1) {
    return (
      this.$(`ul.gear-sub-menu a[data-id="${id}"]`) as ZeptoResult
    ).parent()
  }

  private getTitle() {
    return (this.core.options.levelSelector || {}).title
  }

  private onLevelSwitchStart() {
    this.core.activePlayback.trigger('playback:level:select:start')
    this.levelElement(this.selectedLevelId).addClass('changing')
  }

  private onLevelSwitchEnd() {
    this.levelElement(this.selectedLevelId).removeClass('changing')
  }

  private updateText(level: number) {
    if (level === undefined || isNaN(level)) {
      return
    }
    this.currentText = this.getLevelLabel(level)
    this.deferRender()
  }

  private getLevelLabel(id: number): string {
    if (id === -1) {
      return 'Auto'
    }
    const index = this.levels.findIndex((l) => l.level === id)
    if (index < 0) {
      return 'Auto'
    }
    return this.levelLabels[index] ?? formatLevelLabel(this.levels[index])
  }

  private updateCurrentLevel(info: QualityLevel) {
    trace(`${T} updateCurrentLevel`, { info })
    this.highlightCurrentLevel()
  }

  private highlightCurrentLevel() {
    trace(`${T} highlightCurrentLevel`, {
      selectedLevelId: this.selectedLevelId,
    })
    this.allLevelElements().removeClass('current')
    this.allLevelElements().find('a').removeClass('gcore-skin-active')

    const currentLevelElement = this.levelElement(this.selectedLevelId)

    currentLevelElement.addClass('current')
    currentLevelElement.find('a').addClass('gcore-skin-active')

    this.updateText(this.selectedLevelId)
  }

  private deferRender = debounce(() => this.render(), 0)
}

function formatLevelLabel(level: QualityLevel): string {
  const h = level.width > level.height ? level.height : level.width
  return `${h}p`
}

function debounce(fn: () => void, wait: number) {
  let timerId: ReturnType<typeof setTimeout> | null = null
  return function () {
    if (timerId !== null) {
      clearTimeout(timerId)
    }
    timerId = setTimeout(() => {
      timerId = null
      fn()
    }, wait)
  }
}
