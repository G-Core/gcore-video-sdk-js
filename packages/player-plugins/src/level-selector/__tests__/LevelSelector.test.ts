import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { UICorePlugin } from '@clappr/core'
import FakeTimers from '@sinonjs/fake-timers'
import EventLite from 'event-lite'
import { Logger, LogTracer, setTracer } from '@gcorevideo/utils'
import { LevelSelector } from '../LevelSelector.js'

setTracer(new LogTracer('LevelSelector.test'))
Logger.enable('*')

describe('LevelSelector', () => {
  let clock: FakeTimers.InstalledClock
  beforeEach(() => {
    clock = FakeTimers.install()
  })
  afterEach(() => {
    clock.uninstall()
  })
  describe('options.restrictResolution', () => {
    let core: any
    let levelSelector: LevelSelector
    beforeEach(async () => {
      const activeContainer = Object.assign(new EventLite(), {
        $el: {
          html: vi.fn(),
        },
      })
      const activePlayback = Object.assign(new EventLite(), {
        currentLevel: -1,
        levels: [],
      })
      core = Object.assign(new EventLite(), {
        activeContainer,
        activePlayback,
        options: {
          levelSelector: {
            restrictResolution: 360,
          },
        },
      })
      core.mediaControl = new UICorePlugin(core),
      levelSelector = new LevelSelector(core)
      core.emit('core:ready')
      await clock.tickAsync(1)
      activePlayback.emit('playback:levels:available', [
        {
          level: 0,
          height: 360,
          width: 640,
          bitrate: 300000,
        },
        {
          level: 1,
          height: 720,
          width: 1280,
          bitrate: 150000,
        },
        {
          level: 2,
          height: 1080,
          width: 1920,
          bitrate: 250000,
        },
      ])
      await clock.tickAsync(1)
    })
    it('should render the restricted quality level label', () => {
      expect(levelSelector.el.textContent?.replace('/assets/icons/new/arrow-right.svg', '').replace(/\s+/g, ' ')).toContain('Quality 360p')
      expect(levelSelector.el.innerHTML).toMatchSnapshot()
    })
  })
})
