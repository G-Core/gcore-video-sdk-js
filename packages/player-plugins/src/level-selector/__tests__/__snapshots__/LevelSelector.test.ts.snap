// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LevelSelector > options.restrictResolution > should render the restricted quality level label 1`] = `
"<button class="gplayer-lite-btn gcore-skin-text-color gear-option">
    
    <span>Quality</span>
    <span class="gear-option_arrow-right-icon">/assets/icons/new/arrow-right.svg</span>
    <span class="gear-option_value">360p</span>
</button>
"
`;
